File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/service/impl/TimeServiceImpl.java
package com.github.arhor.simple.expense.tracker.service.impl;

import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.util.TimeZone;

import org.springframework.stereotype.Service;

import com.github.arhor.simple.expense.tracker.service.DateRangeCriteria;
import com.github.arhor.simple.expense.tracker.service.TimeService;
import com.github.arhor.simple.expense.tracker.util.TemporalRange;

import static com.github.arhor.simple.expense.tracker.util.TimeUtil.createDateRange;
import static com.github.arhor.simple.expense.tracker.util.TimeUtil.currentZonedDateTime;

@Service
public class TimeServiceImpl implements TimeService {

    @Override
    public ZonedDateTime now() {
        return now(null);
    }

    @Override
    public ZonedDateTime now(final TimeZone timezone) {
        return currentZonedDateTime(timezone);
    }

    @Override
    public TemporalRange<LocalDate> convertToDateRange(final DateRangeCriteria criteria, final TimeZone timezone) {
        return createDateRange(criteria.startDate(), criteria.endDate(), timezone);
    }
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/service/impl/TimeServiceImpl.kt
package com.github.arhor.simple.expense.tracker.service.impl

import com.github.arhor.simple.expense.tracker.DateRangeCriteria
import com.github.arhor.simple.expense.tracker.service.TimeService
import com.github.arhor.simple.expense.tracker.util.TemporalRange
import com.github.arhor.simple.expense.tracker.util.createDateRange
import com.github.arhor.simple.expense.tracker.util.currentZonedDateTime
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.ZonedDateTime
import java.util.TimeZone

@Service
class TimeServiceImpl : TimeService {

    override fun now(): ZonedDateTime {
        return now(null)
    }

    override fun now(timezone: TimeZone?): ZonedDateTime {
        return currentZonedDateTime(timezone)
    }

    override fun convertToDateRange(criteria: DateRangeCriteria?, timezone: TimeZone?): TemporalRange<LocalDate> {
        return criteria?.let {
            createDateRange(
                it.startDate,
                it.endDate,
                timezone
            )
        } ?: createDateRange(null, null, timezone)
    }
}
